src/models/hybrid.py

```diff
@@ -0,0 +1,289 @@
+        "layer norm" : lambda: nn.LayerNorm(config.hidden_size, eps=config.layer_norm_epsilon),
+        "absolute positional embedding" : lambda: AbsolutePositionalEmbedding(config.max_position_embeddings, config.hidden_size),
+
+        "llama attention"  : _make_llama_attention_factory(config, layer_idx, use_rope=True),
+        "llama attention no rope" : _make_llama_attention_factory(config, layer_idx, use_rope=False),
+        "gpt2 attention"   : lambda: GPT2Attention(config=config, layer_idx=layer_idx),
+        "mamba mixer" : lambda: MambaMixer(config, layer_idx=layer_idx),
+
+        "llama mlp" : lambda: LlamaMLP(config),
+        "gpt2 mlp"  : lambda: GPT2MLP(config.n_inner if config.n_inner is not None else 4 * config.hidden_size, config),
+
+        "llama block" : lambda: LlamaDecoderLayer(config, layer_idx), # pyright: ignore[reportArgumentType]
+        "gpt2 block"  : lambda: GPT2Block(config, layer_idx=layer_idx),
+        "mamba block" : lambda: MambaBlock(config, layer_idx=layer_idx),
+    }
+
+    if not all(key in SUPPORTED_BLOCKS for key in MAPPING.keys()):
+        raise Exception("Not all \"supported\" blocks can be instantiated! Make "
+                        "sure `MAPPING` in this function and `SUPPORTED_BLOCKS` "
+                        "are equal in this file")
+    return MAPPING[spec_name]()
+
+
+class HybridBackbone(nn.Module):
+
+    def has(self, module_substr: str):
+        return any(map(lambda mod: module_substr in mod, self.module_names))
+
+    def __init__(
+        self, 
+        module_names: list[str],
+        n_positions: int,
+        embed_dim: int, 
+        n_layer: int,
+        n_head: int,
+        rope_theta: float = 1e4,
+        **kwargs
+    ):
+        super().__init__()
+
+        flatten = lambda lst: sum(map(flatten, lst), []) if isinstance(lst, list) else [lst]
+        self.module_names = flatten(module_names)
+        if not all(isinstance(mod, str) for mod in self.module_names):
+            raise TypeError(f"Module names for HybridBackbone are malformed! Got:\n{self.module_names}")
+    
+        self.raw_config = {
+            "n_positions" : n_positions,
+            "n_embd" : embed_dim,
+            "n_layer" : n_layer,
+            "n_head" : n_head,
+            "hidden_act" : hidden_act,
+            "rope_theta" : rope_theta,
+            **kwargs
+        }
+
+        self.llama_config = LlamaConfig(
+            max_position_embeddings=2 * n_positions,
+            hidden_size=embed_dim,
+            intermediate_size=4*embed_dim,
+            num_hidden_layers=n_layer,
+            num_attention_heads=n_head,
+            hidden_act=kwargs.get("llama_hidden_act", "silu"),
+            rope_theta=rope_theta,
+            use_cache=False, # On inspection, this only writes to cache, not reads(?)
+            **kwargs # provide all params to only this config to serve as default later on
+        ) if self.has("llama") else None
```

Comment:
Consider the case where we have only modules that don't refer to these three configs. Then we will provide `llama_config` as default, but it will be set to `None` here...

Reply:
Just always instantiate it! It shouldn't cost any significant resources to just instantiate a config